File: tasks/native-runtime/context_engine_parity_probe.py
"""Context-engine parity probe, Python side.

Covers the token engine: `is_cjk`'s ranges through `estimate_tokens`, the message and tool
estimators, the body breakdown, the per-model window lookup, `available_input_tokens`,
`plan_token_budget` and `token_trim` — everything in `context_engine.py` except the
SHA-1-based identity half (`base_context_id`, `build_context_diff`,
`build_engine_diagnostics`), which is deliberately not ported yet.

Settings are patched through a `with_settings` helper because the oracle reads them as
module globals. The small-window cases exist so the trimming path can be exercised without
pushing hundreds of kilobytes through both sides: with `reserve_output_tokens=0`,
`safety_margin_ratio=0.0` and a 40-token window, a handful of messages is already over
budget.

Usage::

    python tasks/native-runtime/context_engine_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example context_engine_parity_probe > ../rust.json
    diff <(tr -d '\\r' < python.json) <(tr -d '\\r' < rust.json)
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

REPO = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO))

from deepseek_infra.infra.gateway import context_engine as ce  # noqa: E402

TEXTS = [
    "",
    "a",
    "hello world",
    "中文",
    "中文 with latin",
    "你好，世界！",  # fullwidth punctuation
    "ひらがなカタカナ",
    "한국어",
    "漢字ＡＢＣ",  # fullwidth latin
    "🙂🙂",
    "a" * 400,
    "中" * 7,
    "mixed 中 and a",
    "ｆｕｌｌ\u3000width",
]

MESSAGES = [
    {},
    {"role": "user", "content": "text"},
    {"role": "user", "content": "中文内容"},
    {"role": "user", "content": [{"type": "text", "text": "中文"}]},
    {"role": "user", "content": [{"type": "image_url", "image_url": {"url": "u"}}]},
    {"role": "user", "content": [{"type": "text"}, {"type": "image_url"}]},
    {"role": "assistant", "content": "ok", "tool_calls": [{"function": {"name": "web_search", "arguments": "{\"q\": \"中文\"}"}}]},
    {"role": "assistant", "tool_calls": ["bad", {"function": "not-a-dict"}]},
    {"role": "user", "content": 5},
]

TOOL_ARRAYS = [
    None,
    [],
    "not-a-list",
    [{"type": "function", "function": {"name": "web_search", "description": "搜索"}}],
    [{"type": "function", "function": {"name": f"tool{i}", "description": "描述"}} for i in range(4)],
]

BODIES = [
    {},
    {"messages": []},
    {"messages": [{"role": "system", "content": "role prompt"}]},
    {"messages": [{"role": "system", "content": "role prompt"}, {"role": "system", "content": "dynamic"}]},
    {"messages": [{"role": "system", "content": "a"}, {"role": "user", "content": "b"}, {"role": "system", "content": "c"}]},
    {"model": "deepseek-v4-pro", "messages": [{"role": "user", "content": "中文"}], "tools": [{"function": {"name": "t"}}]},
]

MODELS = [None, "", "deepseek-v4-pro", "deepseek-v4-flash", "unknown-model", " deepseek-v4-pro "]

# (enabled, token_aware_trim, reserve, safety, threshold, default_window, min_keep, windows)
SETTINGS_CASES = [
    # (enabled, reserve, safety, threshold, default_window, min_keep, windows)
    # `token_aware_trim` is absent on purpose: it belongs to `context_manager`, not the
    # engine, and the engine does not import it.
    (True, 8_192, 0.05, 75.0, 65_536, 2, {"deepseek-v4-pro": 131_072, "deepseek-v4-flash": 131_072}),
    (True, 0, 0.0, 75.0, 40, 2, {"deepseek-v4-pro": 100}),
    (True, 0, 0.0, 75.0, 40, 4, {"deepseek-v4-pro": 100}),
    (True, 8_192, 0.05, 1.0, 65_536, 2, {"deepseek-v4-pro": 131_072}),
    (True, 100_000, 0.05, 75.0, 65_536, 2, {"deepseek-v4-pro": 131_072}),
    (False, 8_192, 0.05, 75.0, 65_536, 2, {"deepseek-v4-pro": 131_072}),
]

TRIM_MESSAGES = [
    {"role": "system", "content": "stable prefix"},
    {"role": "user", "content": "中" * 20},
    {"role": "assistant", "content": "a" * 40},
    {"role": "user", "content": "b" * 40},
    {"role": "system", "content": "dynamic tail"},
]


def with_settings(case, call):
    names = (
        "CONTEXT_ENGINE_ENABLED",
        "CONTEXT_ENGINE_RESERVE_OUTPUT_TOKENS",
        "CONTEXT_ENGINE_SAFETY_MARGIN_RATIO",
        "CONTEXT_ENGINE_COMPRESS_THRESHOLD_PCT",
        "CONTEXT_ENGINE_DEFAULT_CONTEXT_WINDOW",
        "CONTEXT_ENGINE_MIN_KEEP_MESSAGES",
        "CONTEXT_ENGINE_MODEL_CONTEXT_WINDOWS",
    )
    saved = tuple(getattr(ce, name) for name in names)
    for name, value in zip(names, case):
        setattr(ce, name, value)
    try:
        return call()
    finally:
        for name, value in zip(names, saved):
            setattr(ce, name, value)


def main() -> int:
    out: dict[str, object] = {}

    for index, text in enumerate(TEXTS):
        out[f"tokens::{index}"] = ce.estimate_tokens(text)

    for index, message in enumerate(MESSAGES):
        out[f"message::{index}"] = ce.estimate_message_tokens(message)

    for index, tools in enumerate(TOOL_ARRAYS):
        out[f"tools::{index}"] = ce.estimate_tools_tokens(tools)

    for index, body in enumerate(BODIES):
        out[f"breakdown::{index}"] = ce.estimate_body_breakdown(body)

    for index, model in enumerate(MODELS):
        out[f"window::{index}"] = ce.context_window_for_model(model)
        out[f"available::{index}"] = ce.available_input_tokens(model)

    for index, body in enumerate(BODIES):
        out[f"plan::{index}"] = ce.plan_token_budget(body).to_dict()

    for settings_index, case in enumerate(SETTINGS_CASES):
        for model_index, model in enumerate(MODELS):
            out[f"window-s{settings_index}::{model_index}"] = with_settings(
                case, lambda: ce.context_window_for_model(model)
            )
            out[f"available-s{settings_index}::{model_index}"] = with_settings(
                case, lambda: ce.available_input_tokens(model)
            )
        out[f"plan-s{settings_index}"] = with_settings(
            case, lambda: ce.plan_token_budget(BODIES[5]).to_dict()
        )
        trimmed, dropped = with_settings(case, lambda: ce.token_trim(TRIM_MESSAGES, model=None))
        out[f"trim-s{settings_index}::dropped"] = dropped
        out[f"trim-s{settings_index}::kept"] = len(trimmed)
        out[f"trim-s{settings_index}::roles"] = [item.get("role") for item in trimmed]
        out[f"trim-s{settings_index}::first"] = trimmed[0].get("content") if trimmed else None
        out[f"trim-s{settings_index}::last"] = trimmed[-1].get("content") if trimmed else None
        out[f"trim-s{settings_index}::overhead"] = with_settings(
            case, lambda: ce.token_trim(TRIM_MESSAGES, model=None, fixed_overhead_tokens=6)[1]
        )

    out["trim::empty"] = ce.token_trim([], model=None)[0]
    out["trim::empty-dropped"] = ce.token_trim([], model=None)[1]

    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/context_engine_parity_probe.rs
//! Context-engine parity probe, Rust side.
//!
//! Replays the same corpus as `tasks/native-runtime/context_engine_parity_probe.py` through
//! `deepseek_policy::context_engine`.
//!
//! Usage::
//!
//!     python tasks/native-runtime/context_engine_parity_probe.py > python.json
//!     cd rust && cargo run -p deepseek-policy --example context_engine_parity_probe > ../rust.json
//!     diff <(tr -d '\r' < python.json) <(tr -d '\r' < rust.json)

use deepseek_policy::context_engine::{
    ContextEngineSettings, available_input_tokens, base_context_id, build_context_diff,
    build_engine_diagnostics, context_window_for_model, estimate_body_breakdown,
    estimate_message_tokens, estimate_tokens, estimate_tools_tokens, plan_token_budget, token_trim,
};
use serde_json::{Map, Value, json};

/// The oracle's text corpus, in order — the two generated entries included.
fn corpus_texts() -> Vec<String> {
    vec![
        String::new(),
        "a".to_string(),
        "hello world".to_string(),
        "中文".to_string(),
        "中文 with latin".to_string(),
        "你好，世界！".to_string(),
        "ひらがなカタカナ".to_string(),
        "한국어".to_string(),
        "漢字ＡＢＣ".to_string(),
        "🙂🙂".to_string(),
        "a".repeat(400),
        "中".repeat(7),
        "mixed 中 and a".to_string(),
        "ｆｕｌｌ\u{3000}width".to_string(),
    ]
}

fn message_cases() -> Vec<Value> {
    vec![
        json!({}),
        json!({"role": "user", "content": "text"}),
        json!({"role": "user", "content": "中文内容"}),
        json!({"role": "user", "content": [{"type": "text", "text": "中文"}]}),
        json!({"role": "user", "content": [{"type": "image_url", "image_url": {"url": "u"}}]}),
        json!({"role": "user", "content": [{"type": "text"}, {"type": "image_url"}]}),
        json!({"role": "assistant", "content": "ok", "tool_calls": [{"function": {"name": "web_search", "arguments": "{\"q\": \"中文\"}"}}]}),
        json!({"role": "assistant", "tool_calls": ["bad", {"function": "not-a-dict"}]}),
        json!({"role": "user", "content": 5}),
    ]
}

fn tool_arrays() -> Vec<Value> {
    vec![
        Value::Null,
        json!([]),
        json!("not-a-list"),
        json!([{"type": "function", "function": {"name": "web_search", "description": "搜索"}}]),
        json!([
            {"type": "function", "function": {"name": "tool0", "description": "描述"}},
            {"type": "function", "function": {"name": "tool1", "description": "描述"}},
            {"type": "function", "function": {"name": "tool2", "description": "描述"}},
            {"type": "function", "function": {"name": "tool3", "description": "描述"}},
        ]),
    ]
}

fn bodies() -> Vec<Value> {
    vec![
        json!({}),
        json!({"messages": []}),
        json!({"messages": [{"role": "system", "content": "role prompt"}]}),
        json!({"messages": [{"role": "system", "content": "role prompt"}, {"role": "system", "content": "dynamic"}]}),
        json!({"messages": [{"role": "system", "content": "a"}, {"role": "user", "content": "b"}, {"role": "system", "content": "c"}]}),
        json!({"model": "deepseek-v4-pro", "messages": [{"role": "user", "content": "中文"}], "tools": [{"function": {"name": "t"}}]}),
    ]
}

fn models() -> Vec<Option<&'static str>> {
    vec![
        None,
        Some(""),
        Some("deepseek-v4-pro"),
        Some("deepseek-v4-flash"),
        Some("unknown-model"),
        Some(" deepseek-v4-pro "),
    ]
}

fn identity_bodies() -> Vec<Value> {
    let mut all = bodies();
    all.extend(vec![
        json!({"messages": [{"role": "system", "content": "中文前缀"}]}),
        json!({"tools": [{"function": {"name": "a"}}, {"function": {"name": ""}}, {"function": "x"}, "not-a-dict"]}),
        json!({"tools": [{"function": {"name": 5}}]}),
        json!({"messages": [{"role": "user", "content": "x"}], "tools": [{"function": {"name": "b"}}, {"function": {"name": "a"}}]}),
        json!({"messages": [{"role": "user", "content": "x"}], "tools": [{"function": {"name": "a"}}, {"function": {"name": "b"}}]}),
    ]);
    all
}

fn trim_messages() -> Vec<Value> {
    vec![
        json!({"role": "system", "content": "stable prefix"}),
        json!({"role": "user", "content": "中".repeat(20)}),
        json!({"role": "assistant", "content": "a".repeat(40)}),
        json!({"role": "user", "content": "b".repeat(40)}),
        json!({"role": "system", "content": "dynamic tail"}),
    ]
}

fn settings_cases() -> Vec<ContextEngineSettings> {
    let windows = |entries: Vec<(&str, i64)>| {
        entries
            .into_iter()
            .map(|(name, window)| (name.to_string(), window))
            .collect::<Vec<_>>()
    };
    let base = ContextEngineSettings::default();
    vec![
        base.clone(),
        ContextEngineSettings {
            reserve_output_tokens: 0,
            safety_margin_ratio: 0.0,
            default_context_window: 40,
            min_keep_messages: 2,
            model_context_windows: windows(vec![("deepseek-v4-pro", 100)]),
            ..base.clone()
        },
        ContextEngineSettings {
            min_keep_messages: 4,
            reserve_output_tokens: 0,
            safety_margin_ratio: 0.0,
            default_context_window: 40,
            model_context_windows: windows(vec![("deepseek-v4-pro", 100)]),
            ..base.clone()
        },
        ContextEngineSettings {
            compress_threshold_pct: 1.0,
            model_context_windows: windows(vec![("deepseek-v4-pro", 131_072)]),
            ..base.clone()
        },
        ContextEngineSettings {
            reserve_output_tokens: 100_000,
            model_context_windows: windows(vec![("deepseek-v4-pro", 131_072)]),
            ..base.clone()
        },
        ContextEngineSettings {
            enabled: false,
            model_context_windows: windows(vec![("deepseek-v4-pro", 131_072)]),
            ..base
        },
    ]
}

fn main() {
    let mut out: Map<String, Value> = Map::new();
    let default = ContextEngineSettings::default();

    for (index, text) in corpus_texts().iter().enumerate() {
        out.insert(format!("tokens::{index}"), json!(estimate_tokens(text)));
    }

    for (index, message) in message_cases().iter().enumerate() {
        out.insert(
            format!("message::{index}"),
            json!(estimate_message_tokens(message)),
        );
    }

    for (index, tools) in tool_arrays().iter().enumerate() {
        let tools = if tools.is_null() { None } else { Some(tools) };
        out.insert(
            format!("tools::{index}"),
            json!(estimate_tools_tokens(tools)),
        );
    }

    for (index, body) in bodies().iter().enumerate() {
        out.insert(
            format!("breakdown::{index}"),
            estimate_body_breakdown(body).to_value(),
        );
    }

    for (index, model) in models().iter().enumerate() {
        out.insert(
            format!("window::{index}"),
            json!(context_window_for_model(*model, &default)),
        );
        out.insert(
            format!("available::{index}"),
            json!(available_input_tokens(*model, &default)),
        );
    }

    for (index, body) in bodies().iter().enumerate() {
        out.insert(
            format!("plan::{index}"),
            plan_token_budget(body, None, &default).to_value(),
        );
    }

    for (settings_index, settings) in settings_cases().iter().enumerate() {
        for (model_index, model) in models().iter().enumerate() {
            out.insert(
                format!("window-s{settings_index}::{model_index}"),
                json!(context_window_for_model(*model, settings)),
            );
            out.insert(
                format!("available-s{settings_index}::{model_index}"),
                json!(available_input_tokens(*model, settings)),
            );
        }
        out.insert(
            format!("plan-s{settings_index}"),
            plan_token_budget(&bodies()[5], None, settings).to_value(),
        );
        let messages = trim_messages();
        let (trimmed, dropped) = token_trim(&messages, None, 0, settings);
        let roles: Vec<Value> = trimmed
            .iter()
            .map(|item| item.get("role").cloned().unwrap_or(Value::Null))
            .collect();
        out.insert(format!("trim-s{settings_index}::dropped"), json!(dropped));
        out.insert(
            format!("trim-s{settings_index}::kept"),
            json!(trimmed.len()),
        );
        out.insert(format!("trim-s{settings_index}::roles"), json!(roles));
        out.insert(
            format!("trim-s{settings_index}::first"),
            trimmed
                .first()
                .and_then(|item| item.get("content"))
                .cloned()
                .unwrap_or(Value::Null),
        );
        out.insert(
            format!("trim-s{settings_index}::last"),
            trimmed
                .last()
                .and_then(|item| item.get("content"))
                .cloned()
                .unwrap_or(Value::Null),
        );
        out.insert(
            format!("trim-s{settings_index}::overhead"),
            json!(token_trim(&messages, None, 6, settings).1),
        );
    }

    let id_bodies = identity_bodies();
    for (index, body) in id_bodies.iter().enumerate() {
        out.insert(format!("base::{index}"), json!(base_context_id(body)));
        out.insert(format!("diff::{index}"), build_context_diff(body, 0));
        out.insert(
            format!("engine::{index}"),
            build_engine_diagnostics(body, None, 0, &default),
        );
    }
    // Tool order is part of the prefix identity, and the last two bodies differ only there.
    out.insert(
        "base::order-swapped".to_string(),
        json!(base_context_id(&id_bodies[3]) != base_context_id(&id_bodies[4])),
    );
    out.insert(
        "diff::dropped".to_string(),
        build_context_diff(&bodies()[4], 3),
    );
    out.insert(
        "engine-s1".to_string(),
        build_engine_diagnostics(&bodies()[5], None, 0, &settings_cases()[1]),
    );

    let (empty, empty_dropped) = token_trim(&[], None, 0, &default);
    out.insert("trim::empty".to_string(), json!(empty));
    out.insert("trim::empty-dropped".to_string(), json!(empty_dropped));

    let mut encoded =
        serde_json::to_string_pretty(&Value::Object(out)).expect("serialize probe output");
    encoded.push('\n');
    print!("{encoded}");
}
